src/main/java/school/hei/haapi/service/FeeService.java

```diff
@@ -1,135 +1,86 @@
+    Pageable pageable =
+        PageRequest.of(page.getValue() - 1, pageSize.getValue(), Sort.by(DESC, "dueDatetime"));
 
-  public List<Fee> getFeesByStudentId(
-      String studentId, PageFromOne page, BoundedPageSize pageSize,
-      school.hei.haapi.endpoint.rest.model.Fee.StatusEnum status) {
-    Pageable pageable = PageRequest.of(
-        page.getValue() - 1,
-        pageSize.getValue(),
-        Sort.by(DESC, "dueDatetime"));
     if (status != null) {
-      List<Fee> studentFees = resetPaymentRelatedInfo(feeRepository.getByStudentId(studentId));
-      return getFeesByStatus(studentFees, status, page, pageSize);
+      return feeRepository.getFeesByStatus(status, pageable);
     }
-    return resetPaymentRelatedInfo(feeRepository.getByStudentId(studentId, pageable));
+    return feeRepository.getFeesByStatus(DEFAULT_STATUS, pageable);
   }
 
-  private List<Fee> getFeesByStatus(
-      List<Fee> fees,
-      school.hei.haapi.endpoint.rest.model.Fee.StatusEnum status,
-      PageFromOne page, BoundedPageSize pageSize) {
-    int firstIndex = 0;
-    int lastIndex = page.getValue() * pageSize.getValue();
-    if (page.getValue() > 1) {
-      firstIndex = (page.getValue() - 1) * pageSize.getValue();
-    }
-    List<Fee> feesByStatus = fees.stream()
-        .filter(fee -> fee.getStatus().equals(status))
-        .sorted(Comparator.comparing(Fee::getDueDatetime).reversed())
-        .collect(Collectors.toUnmodifiableList());
-    if (firstIndex >= feesByStatus.size()) {
-      return List.of();
-    }
-    if (lastIndex > feesByStatus.size() - 1) {
-      lastIndex = feesByStatus.size();
-    }
-    return feesByStatus.subList(firstIndex, lastIndex);
-  }
-
-  private school.hei.haapi.endpoint.rest.model.Fee.StatusEnum getFeeStatus(Fee fee) {
-    if (computeRemainingAmount(fee) == 0) {
-      return school.hei.haapi.endpoint.rest.model.Fee.StatusEnum.PAID;
-    } else {
-      if (Instant.now().isAfter(fee.getDueDatetime())) {
-        return school.hei.haapi.endpoint.rest.model.Fee.StatusEnum.LATE;
-      }
-      return school.hei.haapi.endpoint.rest.model.Fee.StatusEnum.UNPAID;
+  public List<Fee> getFeesByStudentId(String studentId, PageFromOne page, BoundedPageSize pageSize,
+                                      school.hei.haapi.endpoint.rest.model.Fee.StatusEnum status) {
+    Pageable pageable =
+        PageRequest.of(page.getValue() - 1, pageSize.getValue(), Sort.by(DESC, "dueDatetime"));
+    if (status != null) {
+      return feeRepository.getFeesByStudentIdAndStatus(studentId, status, pageable);
     }
+    return feeRepository.getByStudentId(studentId, pageable);
   }
 
-  private int computeRemainingAmount(Fee fee) {
-    List<Payment> payments = fee.getPayments();
-    if (payments != null) {
-      int amount = payments
-          .stream()
-          .mapToInt(Payment::getAmount)
-          .sum();
-      return fee.getTotalAmount() - amount;
+  public Fee updateFeeStatus(Fee initialFee) {
```

Comment:
Can't you just put this method into the mapper so it's generalized to avoid to call it each time ?

For example during toEntity if it should be persisted after or toDomain if it's calculated.

Reply:
But this method is used in the Scheduler actually...